src/rgw/rgw_rest_s3.cc

```diff
@@ -5612,7 +5616,7 @@ AWSSignerV4::prepare(const DoutPrefixProvider *dpp,
 
   const char* exp_payload_hash = nullptr;
   string payload_hash;
-  if (is_non_s3_op) {
+  if (is_non_s3_op || RGWHandler_REST_PSTopic_AWS::action_exists(info)) {
```

Comment:
i know, it doesn't look great, but is_non_s3_op(RGWOpType op_type) already does that in regular S3 Topic request workflow.
This code is for redirects to master from secondary and in this workflow we do not have is_non_s3_op(RGWOpType op_type).
The `is_non_s3_op` we see here is a bool operator that was passed by the IAM_forward_request and i did not want rewrite another specific `topic_forward_request`.
unless the rgw_rest.cpp can have a generic function that can tell a request is non-s3 based on req_info struct?

Reply:
if there could be a similar function  for IAM t `RGWHandler_REST_PSTopic_AWS::action_exists`, then based on `req_info` struct we could figure if the request is non-s3 and then we do not need to plumb thru the code a BOOL variable

Reply:
> i see, thanks. i had assumed this was on the server side. i think this is ok for now, but please add a `// TODO: teach callers about sns service`
> 
> cc @pritha-srivastava
> 
> we added a lot of special cases for `service == "iam"` in rgw_rest_client.cc to handle the differences in signing for `forward_iam_request_to_master()`. all of the `rgw_forward_request_to_master()` callers are assumed to be s3
> 
> i'm thinking we could add a `const std::string& service = "s3"` default argument to `rgw_forward_request_to_master()` so that these pubsub APIs could pass in "sns" instead
> 
> then we'd have to thread that through `RGWRESTConn::forward()` -> `RGWRESTSimpleRequest::forward_request()`, and change all the `service == "iam"` checks to `service != "s3"`

Add a comment about sns